File: src/test/scala/dregex/SerializerTest.scala
package dregex

import dregex.impl.RegexParser
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.slf4j.LoggerFactory

import java.time.Duration

/**
  * Test the full cycle of serialization and parsing.
  */
class SerializerTest {

  private[this] val logger = LoggerFactory.getLogger(classOf[SerializerTest])

  @Test
  def testSerialize() = {
    val generator = new TreeGenerator
    var i = 0
    val start = System.nanoTime()
    for (tree <- generator.generate(maxDepth = 3)) {
      i += 1
      val canonicalTree = tree.canonical
      logger.trace("Generated tree: " + canonicalTree)
      val serialized = canonicalTree.toRegex
      logger.debug("Serialized: " + serialized)
      val parserRegex = RegexParser.parse(serialized, new RegexParser.Flags())
      val reparsed = parserRegex.getTree.canonical
      logger.trace("Reparsed: " + reparsed)
      val reserialized = reparsed.toRegex
      logger.trace("Reserialized: " + reserialized)
      assertEquals(reparsed, canonicalTree)
      assertEquals(serialized, reserialized)
    }
    val elapsed = Duration.ofNanos(System.nanoTime() - start)
    logger.debug(s"Trees iteration took: $elapsed; size: $i")
  }

}

File: src/test/java/dregex/SerializerTest.java
package dregex;

import static org.junit.jupiter.api.Assertions.assertEquals;

import dregex.impl.RegexParser;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Test the full cycle of serialization and parsing.
 */
class SerializerTest {

    private static final Logger logger = LoggerFactory.getLogger(SerializerTest.class);

    @Test
    void testSerialize() {
        var generator = new TreeGenerator();
        var i = new AtomicInteger();
        var start = System.nanoTime();
        generator.generate(3).forEach(tree -> {
            i.incrementAndGet();
            var canonicalTree = tree.canonical();
            logger.trace("Generated tree: {}", canonicalTree);
            var serialized = canonicalTree.toRegex();
            logger.debug("Serialized: {}", serialized);
            var parserRegex = RegexParser.parse(serialized, new RegexParser.Flags());
            var reparsed = parserRegex.getTree().canonical();
            logger.trace("Reparsed: {}", reparsed);
            var reserialized = reparsed.toRegex();
            logger.trace("Reserialized: {}", reserialized);
            assertEquals(reparsed, canonicalTree);
            assertEquals(serialized, reserialized);
        });
        var elapsed = Duration.ofNanos(System.nanoTime() - start);
        logger.debug("Trees iteration took: {}; size: {}", elapsed, i);
    }
}
